scalafix-core/shared/src/main/scala/scalafix/internal/rule/DisableSyntax.scala

```diff
@@ -35,38 +37,139 @@ final case class DisableSyntax(
             .at(message, pos(matcher.start))
       }
     }
-    val tokensLintMessage =
-      ctx.tree.tokens.collect {
-        case token @ Keyword(keyword) if config.isDisabled(keyword) =>
-          errorCategory
-            .copy(id = s"keywords.$keyword")
-            .at(s"$keyword is disabled", token.pos)
-        case token @ Token.Semicolon() if config.noSemicolons =>
-          error("noSemicolons", token)
-        case token @ Token.Tab() if config.noTabs =>
-          error("noTabs", token)
-        case token @ Token.Xml.Start() if config.noXml =>
-          error("noXml", token)
-      }.toSeq
-    val treeLintMessages =
-      ctx.tree.collect {
-        case t @ mod"+" if config.noCovariantTypes =>
+    regexLintMessages.result()
+  }
+
+  private def checkTokens(ctx: RuleCtx): Seq[LintMessage] = {
+    ctx.tree.tokens.collect {
+      case token @ Keyword(keyword) if config.isDisabled(keyword) =>
+        errorCategory
+          .copy(id = s"keywords.$keyword")
+          .at(s"$keyword is disabled", token.pos)
+      case token @ Token.Semicolon() if config.noSemicolons =>
+        error("noSemicolons", token)
+      case token @ Token.Tab() if config.noTabs =>
+        error("noTabs", token)
+      case token @ Token.Xml.Start() if config.noXml =>
+        error("noXml", token)
+    }
+  }
+
+  private def checkTree(ctx: RuleCtx): Seq[LintMessage] = {
+    object AbstractWithVals {
+      def unapply(t: Tree): Option[List[Defn.Val]] = {
+        val stats = t match {
+          case Defn.Class(mods, _, _, _, templ)
+              if mods.exists(_.is[Mod.Abstract]) =>
+            templ.stats
+          case Defn.Trait(_, _, _, _, templ) => templ.stats
+          case _ => List.empty
+        }
+        val vals = stats.flatMap {
+          case v: Defn.Val => Some(v)
+          case _ => None
+        }
+        if (vals.isEmpty) None else Some(vals)
+      }
+    }
+
+    object WithMethods {
+      def unapply(t: Tree): Option[List[Defn.Def]] = {
+        val stats = t match {
+          case Defn.Class(_, _, _, _, templ) => templ.stats
```

Comment:
Why not match on `Template`? This implementation currently does not handle defs in `object`

Reply:
Mmm, because I didn't know about it :D